lgsm/functions/workshop_core.sh

```diff
@@ -0,0 +1,318 @@
+				core_exit.sh
+			# Two-factor authentication failure
+			elif [ -n "$(grep "Two-factor code mismatch" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: Two-factor authentication failure"
+				fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: Two-factor authentication failure"
+				core_exit.sh
+			# Incorrect Branch password
+			elif [ -n "$(grep "Password check for AppId" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: betapassword is incorrect"
+				fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: betapassword is incorrect"
+				core_exit.sh
+			# Update did not finish.
+			elif [ -n "$(grep "0x402" "${steamcmdlog}" | tail -1)" ] || [ -n "$(grep "0x602" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Update required but not completed - check network"
+				fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Update required but not completed - check network"
+			else
+				fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Unknown error occured"
+				echo -en "Please provide content log to LinuxGSM developers https://linuxgsm.com/steamcmd-error"
+				fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Unknown error occured"
+			fi
+		elif [ "${exitcode}" != "0" ]; then
+			fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Exit code: ${exitcode}"
+			fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Exit code: ${exitcode}"
+		else
+			fn_print_complete_nl "${commandaction} ${selfname}: ${remotelocation}"
+			fn_script_log_pass "${commandaction} ${selfname}: ${remotelocation}"
+		fi
+
+		if [ "${counter}" -gt "10" ]; then
+			fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: Did not complete the download, too many retrys"
+			fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: Did not complete the download, too many retrys"
+			core_exit.sh
+		fi
+	done
+}
+
+fn_workshop_get_list() {
+	workshoplist=($(echo "${workshopmods}" | tr ";" "\n"))
+}
+
+fn_workshop_get_latest_mod_version() {
+	local modid="$1"
+	local serverresp="$(curl -s -d "itemcount=1&publishedfileids[0]=${modid}" "http://api.steampowered.com/ISteamRemoteStorage/GetPublishedFileDetails/v1")"
+	local remupd=
+	if [[ "${serverresp}" =~ \"hcontent_file\":[[:space:]]*([^,]*) ]]; then
+       remupd="${BASH_REMATCH[1]}"
+    fi
+    echo "${remupd}" | tr -d '"'
+}
+
+fn_workshop_get_name_from_steam() {
+	local modid="$1"
+	local serverresp="$(curl -s -d "itemcount=1&publishedfileids[0]=${modid}" "http://api.steampowered.com/ISteamRemoteStorage/GetPublishedFileDetails/v1")"
+	local title=
+	if [[ "${serverresp}" =~ \"title\":[[:space:]]*([^,]*) ]]; then
```

Comment:
Note that REGEX in BASH, while an awesome feature, is slower than glob pattern matching. You might be thinking that glob pattern matching is limited, so you need REGEX. I used to think the same. However, `[[` uses extended globbing out of the box.

For example:

```bash
if [[ "${serverresp}" == *\"title\":*([[:space:]])(*([^,]))* ]]; then
```

Please test first, in-case I missed something. It's almost 5am. This took much longer than I thought it would.

Refer to `extglob` in bash(1) for more information. It's a useful feature which _usually_ needs to be manually enabled with `shopt -s extglob`, with the exception of `[[`. It's not particularly advanced, but it does cater to most of the common use-cases.

Extended globbing is available in at least >= BASH 3.0, but it being included in `[[` is in >= BASH 4.1.